src/ripple/overlay/impl/PeerSet.cpp

```diff
@@ -26,61 +26,74 @@ namespace ripple {
 
 using namespace std::chrono_literals;
 
-class InboundLedger;
-
-// VFALCO NOTE The txnData constructor parameter is a code smell.
-//             It is true if we are the base of a TransactionAcquire,
-//             or false if we are base of InboundLedger. All it does
-//             is change the behavior of the timer depending on the
-//             derived class. Why not just make the timer callback
-//             function pure virtual?
-//
 PeerSet::PeerSet(
     Application& app,
     uint256 const& hash,
     std::chrono::milliseconds interval,
-    clock_type& clock,
     beast::Journal journal)
     : app_(app)
     , m_journal(journal)
-    , m_clock(clock)
     , mHash(hash)
-    , mTimerInterval(interval)
     , mTimeouts(0)
     , mComplete(false)
     , mFailed(false)
     , mProgress(false)
+    , mTimerInterval(interval)
     , mTimer(app_.getIOService())
 {
-    mLastAction = m_clock.now();
     assert((mTimerInterval > 10ms) && (mTimerInterval < 30s));
 }
 
 PeerSet::~PeerSet() = default;
 
-bool
-PeerSet::insert(std::shared_ptr<Peer> const& ptr)
+void
+PeerSet::addPeers(
+    std::size_t limit,
+    std::function<bool(std::shared_ptr<Peer> const&)> hasItem)
 {
-    ScopedLockType sl(mLock);
+    using ScoredPeer = std::pair<int, std::shared_ptr<Peer>>;
 
-    if (!mPeers.insert(ptr->id()).second)
-        return false;
+    auto const& overlay = app_.overlay();
+
+    std::vector<ScoredPeer> pairs;
+    pairs.reserve(overlay.size());
+
+    overlay.foreach([&](auto const& peer) {
+        auto const score = peer->getScore(hasItem(peer));
+        pairs.emplace_back(score, std::move(peer));
+    });
+
+    std::sort(
+        pairs.begin(),
+        pairs.end(),
+        [](ScoredPeer const& lhs, ScoredPeer const& rhs) {
+            return lhs.first > rhs.first;
```

Comment:
Would it be better if we deferred to the default `std::pair` comparison operator? Used naively, it would result in the things being sorted the "wrong" way, but that can be fixed trivially:

```
    // Sort based on scores, high to low:
    std::sort(pairs.rbegin(), pairs.rend());
```